internal/store/driver.go

```diff
@@ -522,33 +522,67 @@ func (d *Driver) calculateIngressLoadBalancerIPStatus(ing *netv1.Ingress, c clie
-	service, err := d.GetServiceV1(backendSvc.Name, namespace)
+	servicePort, err := d.findServicesPort(service, backendSvc.Port)
+	if err != nil {
+		return 0, "", err
+	}
+
+	protocol, err := d.getPortAnnotatedProtocol(service, servicePort.Name)
 	if err != nil {
-		return 0, err
+		return 0, "", err
 	}
 
+	return servicePort.Port, protocol, nil
+}
+
+func (d *Driver) findServicesPort(service *corev1.Service, backendSvcPort netv1.ServiceBackendPort) (*corev1.ServicePort, error) {
 	for _, port := range service.Spec.Ports {
-		if port.Name == backendSvc.Port.Name {
-			d.log.V(3).Info("Found named port for service", "namespace", namespace, "service", service.Name, "port.name", port.Name, "port.number", port.Port)
-			return port.Port, nil
+		if (backendSvcPort.Number > 0 && port.Port == backendSvcPort.Number) || port.Name == backendSvcPort.Name {
+			d.log.V(3).Info("Found matching port for service", service.Name, "namespace", service.Namespace, "port.name", port.Name, "port.number", port.Port)
+			return &port, nil
 		}
 	}
+	return nil, fmt.Errorf("could not find matching port for service %s, backend port %v, name %s", service.Name, backendSvcPort.Number, backendSvcPort.Name)
+}
+
+func (d *Driver) getPortAnnotatedProtocol(service *corev1.Service, portName string) (string, error) {
+	if service.Annotations != nil {
+		annotation := service.Annotations["k8s.ngrok.com/app-protocols"]
+		if annotation != "" {
+			d.log.V(3).Info("Annotated app-protocols found", "annotation", annotation, "namespace", service.Namespace, "service", service.Name, "portName", portName)
+			m := map[string]string{}
+			err := json.Unmarshal([]byte(annotation), &m)
+			if err != nil {
+				return "", fmt.Errorf("Could not parse protocol annotation: '%s' from: %s service: %s", annotation, service.Namespace, service.Name)
+			}
 
-	return 0, fmt.Errorf("could not find named port for service %s", backendSvc.Name)
+			if protocol, ok := m[portName]; ok {
+				d.log.V(3).Info("Found protocol for port name", "protocol", protocol, "namespace", service.Namespace, "service", service.Name)
+				// only allow cases through where we are sure of intent
+				switch upperProto := strings.ToUpper(protocol); upperProto {
+				case "HTTP", "HTTPS":
+					return upperProto, nil
+				default:
+					return "", fmt.Errorf("Unhandled protocol annotation: '%s', must be 'HTTP' or 'HTTPS'. From: %s service: %s", upperProto, service.Namespace, service.Name)
+				}
+			}
+		}
+	}
+	return "HTTP", nil
 }
 
 // Generates a labels map for matching ngrok Routes to Agent Tunnels
-func (d *Driver) backendToLabelMap(namespace, serviceName string, port int32) map[string]string {
+func (d *Driver) backendToLabelMap(namespace, serviceName string, port int32, protocol string) map[string]string {
 	return map[string]string{
 		"k8s.ngrok.com/namespace": namespace,
 		"k8s.ngrok.com/service":   serviceName,
 		"k8s.ngrok.com/port":      strconv.Itoa(int(port)),
+		"k8s.ngrok.com/protocol":  protocol,
```

Comment:
I don't think we need to label the tunnel group backend or the tunnel with the protocol since the protocol is only used for the connection that is managed by the ingress controller to the kubernetes service. The reason we have these labels are that a service backend has a service name and port and the service name can be re-used across namespaces.

I was thinking we could modify the [TunnelSpec](https://github.com/ngrok/kubernetes-ingress-controller/blob/main/api/v1alpha1/tunnel_types.go#L35-L45) to include an optional `Backend` struct that has the protocol and other information that we can use to configure the TLS client if needed (maybe `ServerName`, `InsecureSkipVerify`, etc down the road).

Reply:
Now updated to store the protocol in a `BackendConfig` in the `TunnelSpec`